FEAT: optimized huggingface model support

KutalVolkan opened the pull request:
#### Overview
The goal of this task is to reintroduce support for HuggingFace models within the PyRIT framework after the original `HuggingFaceChat` was removed due to long model loading time. The aim is to enable support for lightweight models that can run efficiently on a normal machine, while also providing flexibility for more complex setups (e.g., cloud-based inference for larger models).

#### Work Completed

1. **Reintroducing HuggingFaceChat Class:**
   - Refactored the existing `HuggingFaceChat` class to create an optimized `HuggingFaceChatTarget` class that aligns with the current `PromptTarget` design standards.

2. **Test Suite for HuggingFaceChatTarget:**
   - Refactored the existing test suite to align with the newly optimized `HuggingFaceChatTarget` class structure.

3. **Jupytext Integration and Notebook Creation:**
   - Created a `.py` file that serves as both a Python script and a Jupyter notebook using Jupytext, demonstrating the `HuggingFaceChatTarget` usage.

#### Next Steps

1. **Optimize Model Loading for Faster Execution:**
   - Investigate smaller or optimized models that load quickly, such as Phi3 or similar lightweight models.

2. **Refactor and Improve Message Parsing:**
   - Simplify the logic for detecting specific strings (e.g., "ASSISTANT") and parsing responses. Evaluate new libraries or approaches that can make this process more robust and efficient.

3. **Enhance Test Coverage and Mocking:**
   - If needed, expand the test suite to cover more edge cases and error scenarios.
  

4. **Future Support for Cloud-Based HuggingFace Models:**
   - Consider the addition of a HuggingFace target capable of interacting with cloud-hosted endpoints for large models, similar to PyRIT's Azure ML Target
   - Current work has focused on optimizing and refactoring the HuggingFaceChatTarget for local model usage, but cloud integration remains a potential future enhancement.

#### Action Items

- [x] Reintroduce `HuggingFaceChatTarget` class for dynamic HuggingFace model support.
- [x] Add tests to cover initialization, model loading, and response generation with efficient mocking strategies.
- [x] Integrate a Jupyter notebook using Jupytext to demonstrate the new target.
- [x] Optimize model loading with lightweight models.
- [x] Refactor message parsing logic for better maintainability.
- [x] Improve test suite with enhanced mocking and edge case coverage.
- [x] Develop support for cloud-based HuggingFace models and endpoints.
- [x] Update and align notebooks with the rest of the PyRIT repository.

#### Contributors and Feedback

This work is based on suggestions from @romanlutz, who outlined potential areas of improvement and future work. The actual code was reused and built upon from @rdheekonda and @rlundeen2's prior implementations.

---
Related Issue: [#347](https://github.com/Azure/PyRIT/issues/347).

romanlutz commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,201 @@
+    def __init__(
+        self,
+        *,
+        model_id: str = "cognitivecomputations/WizardLM-7B-Uncensored",
```
Not sure a default makes sense here. Wdyt? [I know it's copied from the old code]

KutalVolkan replied:
Hello Roman,

I suggest removing the default value for two reasons:

1. The default model can become outdated or deprecated.
2. Requiring `model_id` ensures users make a conscious decision about which model to use.

However, I would prefer you to make the final decision based on your experience :)

romanlutz replied:
Exactly what I was thinking!

KutalVolkan commented:
Hi @romanlutz ,

I wanted to get your thoughts on the following suggestion: separating concerns by moving all model-related management (such as loading with different precisions, quantization, unloading, and saving models) into a dedicated utility submodule e.g. (`model_management/model_utils.py`).

#### Key Benefits of This Approach:

1. **Inference Time Optimization**: By managing model precision (e.g., `fp16`, `bf16`, `int8`) and quantization separately, we can dynamically adjust model loading based on available hardware and desired trade-offs between performance and memory usage. For instance, quantizing a model from `fp32` to `int8` can significantly reduce memory footprint and speed up inference, which is crucial for faster response times.

2. **Separation of Concerns**: Centralizing model loading, precision handling, quantization, unloading, and saving in a utility module keeps `HuggingFaceChatTarget` focused on its core task—managing prompt interactions—while offloading model-related complexities to a dedicated area.

3. **Maintainability and Reusability**: This approach would make our code more maintainable and reusable. The utility functions can be leveraged by other targets or components, which is particularly useful if we decide to support more models or different configurations.

4. **Flexibility for Use Cases**: The proposed structure allows for flexibility in handling specific use cases, such as changing a model’s precision on the fly based on user requirements (e.g., switching from `fp32` to `int8` for lighter setups) or optimizing for GPU vs. CPU inference scenarios.

I would love to hear your thoughts on whether this aligns with our scope and if there are any additional considerations to keep in mind. In the meantime, I will work on refactoring the message parsing logic and adding support for cloud-based HuggingFace models.

romanlutz commented:
Absolutely! In fact, you've thought about it a lot more thoroughly than I had so far 🙂 I guess I'd do this in chunks rather than one massive PR but I'm open to options. Wdyt?

@rdheekonda and @rlundeen2 have faced issues with the old target and may have thoughts, too.

I'm on the fence about creating a new module for this or just a folder (sub module?) within targets or common. I guess it doesn't really matter and users shouldn't have to interact with it so we should be able to move at a later point if necessary.

KutalVolkan commented:
> Absolutely! In fact, you've thought about it a lot more thoroughly than I had so far 🙂 I guess I'd do this in chunks rather than one massive PR but I'm open to options. Wdyt?

Hi Roman,

Thanks for the feedback! I agree with breaking this into smaller chunks rather than one big PR.

KutalVolkan commented:
Hello @romanlutz,

I made some quick changes by adding `HuggingFaceEndpointTarget`. The endpoint is more or less working, but I’m thinking of removing complete_chat—what are your thoughts on that? It’s a bit rough due to fast prototyping, but I’ll clean it all up in the next two days.

The local version of `HuggingFaceChatTarget` isn’t finished yet and will definitely take longer. It’s exciting to try optimizing it, but I'm currently running it on CUDA, and it takes about 1 minute for microsoft/Phi-3-mini-128k-instruct, so I might not have set something right.

Feedback and ideas are always welcome!

Thanks for your patience :)

dlmgary commented on pyrit/prompt_target/hugging_face_enpoint_target.py:
```diff
@@ -0,0 +1,143 @@
+
+    async def send_prompt_async(self, *, prompt_request: PromptRequestResponse) -> PromptRequestResponse:
+        """
+        Sends a normalized prompt asynchronously to a cloud-based HuggingFace model endpoint.
```
nit: use Google-style comments.

romanlutz commented:
Yes, whatever makes sense. A lot of internal methods there may be "legacy" and can be updated.

@rdheekonda mentioned that we had issues with huge delays while pulling the models from HF and running locally. Even small models took a while. We don't want it to be prohibitively difficult to run the notebooks, for example, so that's something that needs solving....

For tests, that part could probably be mocked so that there's no network call.

KutalVolkan commented:
Hello @romanlutz ,

Here is the workflow I will follow for now:

### 1. Profile the Model Loading Workflow

I will use `cProfile` and other timing  and profiling tools to analyze the entire model loading workflow, covering:

- **Internet to Disk**: Measure the time it takes to download models from Hugging Face to the local disk. This will help understand network-related delays.
- **Disk to Memory**: Measure the time taken to load models from disk into memory.
- **Optional Model Initialization and Inference**: Assess the initialization time and inference speed, which are important for understanding how models perform under different conditions (e.g., CPU vs. GPU).

### 2. Experiment with Class-Level Caching

To optimize the model loading process, I will implement **class-level caching**. This caching mechanism will store the model and tokenizer in memory after the first load, hopefully allowing subsequent instances of `HuggingFaceChatTarget` to reuse the cached objects instead of reloading them from disk. This could help reduce redundant loading times. I will test the impact of this approach to determine its effectiveness in practice.

### 3. Extend and Enhance Test Coverage

While profiling will give us performance data, I plan as you suggested to enhance the test coverage to ensure robustness and proper error handling. This includes:

- **Mocking Internet-to-Disk and Disk-to-Memory Phases**: To simulate and test different conditions without actual downloads, such as slow network conditions, network unavailability, or disk I/O issues.
- **Error Handling for Network Issues**: Mocking scenarios where network errors occur to validate that our code handles these gracefully.

Let me know if there are any specific areas you’d like me to focus on.

romanlutz commented:
All these make sense!

Good point about caching. If the download is a one-time operation of a few minutes it's perhaps fine. If it takes 10 minutes every time that's not great 😄

KutalVolkan commented:
Hello @romanlutz,

You can run the notebook `use_huggingface_chat_target.ipynb` in **Anaconda Prompt as Administrator** to avoid permission issues and improve download efficiency.

### My Observations:

- **Without Admin**: It took around **529 seconds** due to redundant file handling.
- **With Admin**: It only took **0.66 seconds** for downloading specific files and about **52 seconds** for loading the model.

Note: The **0.66 seconds** download time indicates that specific files were downloaded, but I may not have provided all the necessary files. With all required files specified, the entire process could potentially be reduced to under **1 minute**.

To optimize the process, ensure to provide the `necessary_files` list to avoid downloading the entire model and speed up the setup.

There may still be additional bottlenecks, and the code is a bit messy :), but the main problem with downloading should be solved by running as admin.

Performance Data Snippet:
```
Error during table creation: (duckdb.duckdb.IOException) IO Error: File is already open in 
C:\Users\vkuta\anaconda3\envs\pyrit-dev\python.exe (PID 22208)
(Background on this error at: https://sqlalche.me/e/20/e3q8)
Model microsoft/Phi-3-mini-4k-instruct not fully found in cache. Downloading missing files using CLI...
Please make sure you are logged in to Hugging Face CLI. Running login process...
Token is valid (permission: write).
Your token has been saved in your configured git credential helpers (manager).
Your token has been saved to C:\Users\vkuta\.cache\huggingface\token
Login successful

Successfully logged into Hugging Face.
The following files will be downloaded:
  - config.json
  - generation_config.json
  - model-00001-of-00002.safetensors
  - model-00002-of-00002.safetensors
  - special_tokens_map.json
  - tokenizer.json
  - tokenizer.model
  - tokenizer_config.json
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.
Downloaded files:
Specified files for microsoft/Phi-3-mini-4k-instruct downloaded successfully to the cache.
Download of specific files completed in 0.66 seconds.
Loading checkpoint shards: 100%
 2/2 [00:49<00:00, 25.32s/it]
         1097875 function calls (1044540 primitive calls) in 52.512 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        2    0.000    0.000   52.511   26.256 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\IPython\core\interactiveshell.py:3541(run_code)
   1008/2    0.021    0.000   52.511   26.256 {built-in method builtins.exec}
        1    0.000    0.000   52.511   52.511 C:\Users\vkuta\projects\PyRIT\pyrit\prompt_target\hugging_face_chat_target.py:28(__init__)
        1    0.000    0.000   52.286   52.286 C:\Users\vkuta\projects\PyRIT\pyrit\prompt_target\hugging_face_chat_target.py:68(load_model_and_tokenizer)
        1    0.000    0.000   50.149   50.149 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\transformers\models\auto\auto_factory.py:444(from_pretrained)
        1    0.003    0.003   49.382   49.382 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\transformers\modeling_utils.py:2746(from_pretrained)
        1    0.142    0.142   49.162   49.162 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\transformers\modeling_utils.py:3846(_load_pretrained_model)
        2    0.005    0.003   46.857   23.428 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\transformers\modeling_utils.py:650(_load_state_dict_into_model)
    846/2    0.035    0.000   46.852   23.426 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\transformers\modeling_utils.py:676(load)
      299    0.089    0.000   46.779    0.156 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\torch\nn\modules\module.py:1996(_load_from_state_dict)
      195   46.614    0.239   46.614    0.239 {method 'copy_' of 'torch._C.TensorBase' objects}
        2    2.017    1.008    2.035    1.017 {built-in method gc.collect}
        1    0.000    0.000    1.953    1.953 C:\Users\vkuta\projects\PyRIT\pyrit\common\huggingface_model_manager.py:90(download_specific_files_with_cli)
        2    0.000    0.000    1.725    0.862 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\subprocess.py:506(run)
       98    1.701    0.017    1.701    0.017 {method 'acquire' of '_thread.lock' objects}
        2    0.000    0.000    1.699    0.849 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\subprocess.py:1165(communicate)
        2    0.000    0.000    1.699    0.849 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\subprocess.py:1603(_communicate)
        4    0.000    0.000    1.696    0.424 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\threading.py:1087(join)
       44    0.000    0.000    1.696    0.039 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\threading.py:1125(_wait_for_tstate_lock)
        1    0.000    0.000    1.062    1.062 C:\Users\vkuta\projects\PyRIT\pyrit\common\huggingface_model_manager.py:16(login_to_huggingface)
  871/180    0.004    0.000    0.740    0.004 <frozen importlib._bootstrap>:1165(_find_and_load)
  203/178    0.000    0.000    0.740    0.004 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\importlib\__init__.py:108(import_module)
  203/178    0.000    0.000    0.739    0.004 <frozen importlib._bootstrap>:1192(_gcd_import)
   755/83    0.003    0.000    0.738    0.009 <frozen importlib._bootstrap>:1120(_find_and_load_unlocked)
   690/82    0.002    0.000    0.721    0.009 <frozen importlib._bootstrap>:666(_load_unlocked)
   686/82    0.001    0.000    0.719    0.009 <frozen importlib._bootstrap_external>:934(exec_module)
    32/16    0.000    0.000    0.705    0.044 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\huggingface_hub\utils\_validators.py:98(_inner_fn)
 1551/165    0.001    0.000    0.692    0.004 <frozen importlib._bootstrap>:233(_call_with_frames_removed)
        4    0.000    0.000    0.684    0.171 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\requests\sessions.py:500(request)
        4    0.000    0.000    0.674    0.169 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\requests\sessions.py:673(send)
        4    0.000    0.000    0.673    0.168 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\huggingface_hub\utils\_http.py:63(send)
        4    0.000    0.000    0.673    0.168 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\requests\adapters.py:613(send)
        4    0.000    0.000    0.670    0.168 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\urllib3\connectionpool.py:594(urlopen)
        4    0.000    0.000    0.669    0.167 C:\Users\vkuta\anaconda3\envs\pyrit-dev\Lib\site-packages\urllib3\connectionpool.py:379(_make_request)
27857/6489    0.007    0.000    0.604    0.000 {built-in method builtins.hasattr}
   117/94    0.000    0.000    0.603    0.006 C:\Users\vkuta\anaconda3\envs\pyrit-de
   ....
```

romanlutz commented:
Woah! That's a massive difference. I'll try this out and will report my findings. Might take me until tomorrow, though.

My first thought was that perhaps it did cache. The difference is so huge.

KutalVolkan commented:
> Woah! That's a massive difference. I'll try this out and will report my findings. Might take me until tomorrow, though.
> 
> My first thought was that perhaps it did cache. The difference is so huge.

Hopefully i did not mistake, i was quite tired and in a bubble, but let me know foe sure i am excited about your findings and i will work tomorrow further on it and let you know if i messed up :)

> Woah! That's a massive difference. I'll try this out and will report my findings. Might take me until tomorrow, though.
> 
> My first thought was that perhaps it did cache. The difference is so huge.

Unfortunately, your thought was right. I deleted the cache and ran it again, and it downloaded model-00001-of-00002.safetensors, etc., again even though they were downloaded before.

I will play around with this tomorrow.
<img width="536" alt="image" src="https://github.com/user-attachments/assets/b7d38fce-3e00-4965-bf7c-1d65ca691590">

KutalVolkan commented:
Hello @romanlutz ,

I’ve been employing different methods to download the `Phi-3-Mini-4K-Instruct` model. Here are the details of the setups and their performance:

1. **Hugging Face CLI with Parallel Downloads**:
   - **Setting**: Enabled parallel downloads by setting `HF_HUB_ENABLE_HF_TRANSFER=1`.
   - **model-00001-of-00002.safetensors**:
     - **File Size**: 4.97 GB
     - **Download Time**: 101 seconds
     - **Calculated Speed**: Approximately 50.43 MB/s
   - **model-00002-of-00002.safetensors**:
     - **File Size**: 2.67 GB
     - **Download Time**: 52 seconds
     - **Calculated Speed**: Approximately 52.58 MB/s

2. **Aria2 Setup**:
   - I also experimented with aria2c to see if there could be an improvement in download speeds using its capabilities to handle multiple connections. However, the speeds observed were similar to those achieved with the Hugging Face CLI setup.

3. **Default Setup**:
   - Additionally, I tested the standard method provided by the transformers library without any specific configurations for parallel transfers, resulting in comparable speeds to the above methods.
   
My best download time was  `2 min and 14 sec `, with another instance taking  `2 min and 42 sec `. These speeds can vary based on the time of day, such as early morning or late evening, depending on how heavily the servers are loaded.

Despite these diverse approaches, my experiences have been quite consistent across the board. My internet connection offers up to  `381.05 Mbps (~47.6 MB/s) `, and at times, the download speeds slightly exceed this limit, possibly due to ISP speed bursts. This performance suggests that the HF CLI setup with parallel downloads enabled appears to be optimal among the methods I've tested, efficiently utilizing my internet capacity.

I am now shifting my focus to work on the inference part and will revisit the download strategies later. 

It would be helpful to hear about your experiences or from other team members (@rdheekonda, @rlundeen2), particularly any download speeds or feedback you could share. If you've conducted any tests or have insights into optimizing these downloads further, your input would be invaluable.

---
For more information about optimizing download speeds using the Hugging Face CLI and enabling `hf_transfer` for faster transfers, check out the [environment variables guide](https://huggingface.co/docs/huggingface_hub/v0.24.6/package_reference/environment_variables#hfhubenablehftransfer) and the [CLI download guide](https://huggingface.co/docs/huggingface_hub/guides/download).

To calculate the download speed, i used the formula:
```
Speed (MB/s) = File Size (MB) / Download Time (seconds)
```

For example:
- **File Size**: 4.97 GB = 4.97 * 1024 MB = 5093.28 MB
- **Download Time**: 101 seconds

```
Speed = 5093.28 MB / 101 s ≈ 50.43 MB/s
```

Best,
Volkan 😄

rdheekonda commented:
Thank you for the details, @KutalVolkan . While downloading is one aspect during PyRIT demos testing, HuggingFace typically checks the default cache directory, and since we'll be using the same model for demos testing, it should already exist. 

However, our main focus for this feature is on speeding up inference by sending a few text prompts and observing the response in CPU based compute where we run the demos before creating a PR to ensure existing functionality is not broken with new features. Could you test with a few smaller models and check the inference time using the `PromptSendingOrchestrator`? 

The models don’t have to be from Microsoft, as HuggingFace hosts models from various companies. Here’s another 135M [model ](https://huggingface.co/HuggingFaceTB/SmolLM-135M-Instruct/tree/main)with a 269MB size, which I believe will have faster inference compared to phi-3-4k-instruct. 

Again, the goal is to optimize the notebook demo with a small yet effective instruction model so that users can experiment with different HuggingFace models.

KutalVolkan commented:
> Thank you for the details, @KutalVolkan . While downloading is one aspect during PyRIT demos testing, HuggingFace typically checks the default cache directory, and since we'll be using the same model for demos testing, it should already exist.
> 
> However, our main focus for this feature is on speeding up inference by sending a few text prompts and observing the response in CPU based compute where we run the demos before creating a PR to ensure existing functionality is not broken with new features. Could you test with a few smaller models and check the inference time using the `PromptSendingOrchestrator`?
> 
> The models don’t have to be from Microsoft, as HuggingFace hosts models from various companies. Here’s another 135M [model ](https://huggingface.co/HuggingFaceTB/SmolLM-135M-Instruct/tree/main)with a 269MB size, which I believe will have faster inference compared to phi-3-4k-instruct.
> 
> Again, the goal is to optimize the notebook demo with a small yet effective instruction model so that users can experiment with different HuggingFace models.

Hello @rdheekonda ,

Thank you for the feedback!

I have tested the inference times for  two smaller models using the `PromptSendingOrchestrator` on CPU-based compute. Here are the results:

1. **Model: `microsoft/Phi-3-mini-4k-instruct`**  
   - **Prompts Used**: ["What is 3x3?", "What is 4x4?", "What is 5x5?", "What is 6x6?", "What is 7x7?"]  
   - **Inference Time**: Approximately **25.7 seconds**  
   - **Average Inference Time per Prompt**: Approximately **5.14 seconds**  

2. **Model: `HuggingFaceTB/SmolLM-135M-Instruct`**  
   - **Prompts Used**: ["What is 3x3?", "What is 4x4?", "What is 5x5?", "What is 6x6?", "What is 7x7?"]  
   - **Inference Time**: Approximately **12.8 seconds**  
   - **Average Inference Time per Prompt**: Approximately **2.56 seconds**  

Both models were run on CPU, but our current implementation also supports running on GPU.

**Note**: Increasing the `max_new_tokens` parameter leads to an increase in inference time. This is something to consider when optimizing for speed in different scenarios.

Moving forward, I'll be cleaning up the code and considering ways to extract only the assistant's response. A straightforward method would be removing the initial prompt from the output, while a more precise approach would involve using the `tokenizer_config.json` or `tokenizer.chat_template` to identify the LLM response correctly.

Please let me know if further testing is needed or if there are additional models you'd like me to evaluate.

KutalVolkan commented:
Hello @romanlutz,

You can check the detailed notebook for **Model Performance and Response Times** [here](https://github.com/Azure/PyRIT/blob/9ae3d890a10835925603c6a6c026eb826bae1159/doc/code/orchestrators/use_huggingface_chat_target.ipynb).

The tests were conducted using a CPU, and the following are the average response times for each model:

- `microsoft/Phi-3-mini-4k-instruct`: **4.89 seconds**
- `HuggingFaceTB/SmolLM-135M-Instruct`: **3.09 seconds**
- `HuggingFaceTB/SmolLM-360M-Instruct`: **3.31 seconds**
- `HuggingFaceTB/SmolLM-1.7B-Instruct`: **5.87 seconds**
- `Qwen/Qwen2-0.5B-Instruct`: **1.38 seconds**
- `Qwen/Qwen2-1.5B-Instruct`: **2.96 seconds**
- `stabilityai/stablelm-2-zephyr-1_6b`: **5.31 seconds**
- `stabilityai/stablelm-zephyr-3b`: **8.37 seconds**
- `google/gemma-2b`: **7.79 seconds**
- `TinyLlama/TinyLlama-1.1B-Chat-v1.0`: **4.03 seconds**
- `princeton-nlp/Sheared-LLaMA-1.3B-ShareGPT`: **3.36 seconds**
- `princeton-nlp/Sheared-LLaMA-2.7B-ShareGPT`: **8.29 seconds**

I will start cleaning up the code and further play on handling the assistant's response. For now, it might be best to keep the current approach regarding the assistant's output or to extract the initial prompt from the response, which contains both the initial prompt and the assistant's response.

Another approach could involve using markers from `tokenizer_config.json` or working with `tokenizer.chat_template`. However, this felt a bit overcomplicated due to the varying markers; it seemed like it required maintaining a list of all possible markers for an assistant and checking against that.

romanlutz commented:
Those times are pretty good! Thank you for getting to the bottom of this. Looks like it's reasonable across the board with various models.

I haven't used this before so having the response include the prompt sounds odd but if you can extract it out as you wrote that seems... fine (?)

KutalVolkan commented:
Hello @romanlutz,
 
**Summary of Changes**:
- **Prompt Formatting**: Introduced the `format_prompt` method to format prompts for models without a `chat_template`. For models that support a `chat_template`, the tokenizer applies it directly, while for others, a fallback format based on the tokenizer configuration is used.
  - **Current Focus**: Handling models with a `chat_template`. For models without one, users can extend the `format_prompt` method in `pyrit\common\dynamic_prompt_formatter.py`. This will be mentioned in the notebook to guide users toward making necessary changes.
  
- **Robust Method for Assistant Response Extraction**: Implemented a method to extract the assistant's response by slicing generated output tokens after the input tokens. This approach has been tested and works well for most models.

**Known Issue**:
- **Models Producing Garbage Output**: Certain models, such as `google/gemma-2b`, `princeton-nlp/Sheared-LLaMA-1.3B-ShareGPT`, and `princeton-nlp/Sheared-LLaMA-2.7B-ShareGPT`, generate incorrect or incomplete responses. This might result from improper prompt formatting or inherent model behavior. The issue persists even when using a generic pipeline like:
```python
# Use a pipeline as a high-level helper
from transformers import pipeline

# Initialize the pipeline with the desired model
pipe = pipeline("text-generation", model="google/gemma-2b")

# Input text to test with
input_text = "What is 7 * 7?"
result = pipe(input_text, max_new_tokens=50, do_sample=True)

# Print the generated result
print(f"Generated text from: {result}")
```

**Proposed Solution**:
- **For now**, I will handle models with a `chat_template`. Users needing to extend support for models without a `chat_template` can do so by modifying the `format_prompt` method in `pyrit\common\dynamic_prompt_formatter.py`. I will provide one example to guide users.

**Pros**:
- **Scalability and Flexibility**: The approach is easily adaptable for new models and tokenizer configurations.
- **Customization**: By allowing users to modify the `format_prompt` method, we offer flexibility to handle various tokenization needs.

**Challenges**:
- **Manual Updates**:  For models without a `chat_template`, users will need to manually extend the `format_prompt` method to handle specific tokenization needs. This approach minimizes ongoing maintenance on our side but requires users to adapt the `format_prompt` method when using models without a `chat_template`. Users will need to handle these cases by updating the method as they integrate such models or as model configurations evolve.

---
What are your thoughts on this overall approach? I'd appreciate any feedback or suggestions you might have.

rdheekonda commented:
Thanks for the details and working on this, @KutalVolkan. Applying the chat template seems like a good first solution and shown promising during the initial implementation when I started working on this. I believe it works well with instruct models. 

In the early stages, we can likely disregard non-instruct models (ex., google/gemma-2b, princeton-nlp/Sheared-LLaMA-1.3B-ShareGPT, and princeton-nlp/Sheared-LLaMA-2.7B-ShareGPT). With instruct models (`HuggingFaceTB/SmolLM-135M-Instruct`), we get a clear data structure—a list of dictionaries with user entry followed by assistant entry. 

For the Hugging Face implementation, I would skip the models you've mentioned (e.g., gemma-2b) since they aren't instruct models. It's difficult to parse exact assistant responses, especially when they don't follow a chat template. We could document these details so users will be aware.

KutalVolkan commented:
Thanks for the feedback, @rdheekonda. I've implemented the changes based on your suggestions, and the updates are now ready for review. Feel free to provide any additional feedback or suggestions.

Regarding `aria2`, it's still in place for now, but since we can handle transfers directly via `huggingface-cli` with `HF_HUB_ENABLE_HF_TRANSFER=1`, we can consider removing it if you think it's no longer necessary. Let me know your thoughts on that or if there are other components we might not need.

Looking forward to your feedback :)

rdheekonda commented:
Thanks for your contribution, it looks great!

At a high level:
1. Could you please remove the demo for the Hugging Face Chat endpoint demo and keep only local target demo since we don't plan to include it in the testing. Similar to the Ollama functionality, where it exists but has no demo.
2. Also, could you consider mocking the unit tests instead of calling the sample model directly. I'll review again once these changes are made.

KutalVolkan commented:
> Thanks for your contribution, it looks great!
> 
> At a high level:
> 
> 1. Could you please remove the demo for the Hugging Face Chat endpoint demo and keep only local target demo since we don't plan to include it in the testing. Similar to the Ollama functionality, where it exists but has no demo.
> 2. Also, could you consider mocking the unit tests instead of calling the sample model directly. I'll review again once these changes are made.
> 3. Update env_var names in .env_example

Hello @rdheekonda,

The changes have been made and are ready for review. Please feel free to provide any feedback; it's welcome and appreciated.

romanlutz commented on doc/code/orchestrators/harmbench_testing.py:
Why is this in here? Or do you just have to pull main?

KutalVolkan replied:
I just have to pull main.

romanlutz commented on pyrit/common/download_hf_model_with_aria2.py:
```diff
@@ -0,0 +1,89 @@
+from huggingface_hub import HfApi
+
+# Define the base directory for the project
+PYRIT_PATH = Path(__file__, "..", "..").resolve()
```
I think we have this defined in common.paths

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+    "stabilityai/stablelm-2-zephyr-1_6b",
+    "stabilityai/stablelm-zephyr-3b",
+]
+
```
I'd suggest sending one model (ex., "HuggingFaceTB/SmolLM-135M-Instruct") so that demo run time will be minimized.

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+]
+
+# List of prompts to send
+prompt_list = ["What is 3*3?", "What is 4*4?", "What is 5*5?", "What is 6*6?", "What is 7*7?"]
```
Could we consider having just 1-2 prompts, having more prompts might lead to more time to run the demo.

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+# Dictionary to store average response times
+model_times = {}
+
+for model_id in model_list:
```
We could remove iterating through all the models

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+
+        # Print the conversations
+        orchestrator.print_conversations()
+        print("-" * 50)
```
This could be removed once we remove the for loop

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+        orchestrator.print_conversations()
+        print("-" * 50)
+
+    except Exception as e:
+        print(f"An error occurred with model {model_id}: {e}\n")
+        model_times[model_id] = None
+        continue
```
Once the above refactor is done, the exception handling could be updated for a single model

rdheekonda commented on doc/code/orchestrators/use_huggingface_chat_target.py:
```diff
@@ -0,0 +1,108 @@
+    if avg_time is not None:
+        print(f"{model_id}: {avg_time:.2f} seconds")
+    else:
+        print(f"{model_id}: Error occurred, no average time calculated.")
```
Once the above refactoring is complete, could we move this logic for a single model.

rdheekonda commented on pyrit/common/download_hf_model_with_aria2.py:
```diff
@@ -0,0 +1,87 @@
+
+# Define the new folder for Hugging Face models within the same hierarchy as 'pyrit'
+HF_MODELS_DIR = PYRIT_PATH / "hf_models"
+HUGGINGFACE_TOKEN = os.getenv("HUGGINGFACE_TOKEN")
+
+logger = logging.getLogger(__name__)
+
+if not HUGGINGFACE_TOKEN:
+    raise ValueError("HUGGINGFACE_TOKEN environment variable is not set. Please set it before running this function.")
```
Could we move this logic of obtaining hugging face token to the target following code [here](https://github.com/Azure/PyRIT/blob/f4f2b92da1e66e0984cf0c6c027aa267a79f8d6d/pyrit/prompt_target/prompt_chat_target/azure_openai_gptv_chat_target.py#L126)

rdheekonda commented on pyrit/common/download_hf_model_with_aria2.py:
```diff
@@ -0,0 +1,87 @@
+    api = HfApi()
+    try:
+        model_info = api.model_info(model_id, token=HUGGINGFACE_TOKEN)
+        available_files = [file.rfilename for file in model_info.siblings]
```
Could we do a simple validation something like `if not len(available_files): raise ValueError`?

rdheekonda commented on pyrit/common/download_hf_model_with_aria2.py:
```diff
@@ -0,0 +1,87 @@
+        "aria2c",
+        "-d",
+        download_dir_str,
+        "-x",
+        "3",
+        "-s",
+        "5",
+        "-j",
+        "4",
```
could you add a comment explaining the param values -x 3, -s 5, -j 4?

rdheekonda commented on pyrit/common/download_hf_model_with_hf_cli.py:
```diff
@@ -0,0 +1,118 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import os
+import logging
+import subprocess
+from pyrit.common import default_values
+from huggingface_hub import HfApi  # Hugging Face API for model size
```
Nit: order of imports

rdheekonda commented on pyrit/common/download_hf_model_with_hf_cli.py:
```diff
@@ -0,0 +1,118 @@
+    except subprocess.CalledProcessError as e:
+        logger.info(f"Error downloading specific files for model {model_id}: {e}")
+        logger.info(e.stderr)
+        raise
```
Why use a CLI-based download when we already have aria2? Will it provide any faster downloads? If there's no significant improvement in speed, it would make sense to remove the CLI option and rely solely on aria2 for faster and more efficient downloads.

I think `aria2` is optimized for parallel downloads and connection handling, which may outperform basic CLI download tools. If aria2 consistently offers better performance or reliability, removing the CLI option would simplify the process without sacrificing download speed.

KutalVolkan replied:
You're right! After re-testing, the speed difference between Hugging Face CLI with `HF_HUB_ENABLE_HF_TRANSFER=1` and `aria2` is minimal. However, `aria2` offers more flexibility and advanced features like segmented downloads and better control. Given this, it makes sense to simplify the process and focus on using `aria2` for all downloads, as it provides similar speed with more customization options.

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+logger = logging.getLogger(__name__)
+
+
+class HuggingFaceChatTarget(PromptTarget):
```
Since these are instruct models, we could configure the system prompt, could we inherit `PromptChatTarget`?

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+        necessary_files: list = None,
+        max_new_tokens: int = 20,  # Default max_new_tokens parameter
+        temperature: float = 1.0,  # Default temperature parameter
+        top_p: float = 1.0,  # Considers all possible tokens (100% of the probability distribution)
```
Nit: could we remove the comments from here, those are self-explanatory :)

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+            if self.necessary_files is None:
+                # Perform general download if no specific files are mentioned
+                logger.info(f"Downloading the entire model {self.model_id} since no specific files are provided...")
+                download_model_with_cli(self.model_id)
```
If aria2 is faster, could we use that option by default?

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+            # Load the tokenizer and model from the specified directory
+            logger.info(f"Loading model {self.model_id} from cache path: {cache_dir}...")
+            self.tokenizer = AutoTokenizer.from_pretrained(self.model_id, cache_dir=cache_dir)
+            # self.model = AutoModelForCausalLM.from_pretrained(self.model_id, device_map="auto", cache_dir=cache_dir)
```
Nit: remove comment

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+                "with a chat template."
+            )
+            logger.error(error_message)
+            raise ValueError(error_message)
```
There's a lot happening in this function. Could we refactor it by extracting `apply_chat_template` into a new instance method and make it private?

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+            generated_tokens = generated_ids[0][input_length:]
+
+            # Decode the assistant's response from the generated token IDs
+            assistant_response = self.tokenizer.decode(generated_tokens, skip_special_tokens=True).strip()
```
Instead of hardcoding skip_special_tokens=True in the decode method, we can introduce skip_special_tokens as an instance variable during initialization. This way, users can pass a boolean value to control whether special tokens should be skipped or not.

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,241 @@
+
+            logger.info(f"Assistant's response: {assistant_response}")
+
+            prompt_response = construct_response_from_request(
```
If the response is empty, could we implement a retry mechanism similar to other targets for improved resilience?

rdheekonda commented on tests/test_hf_model_downloads.py:
```diff
@@ -0,0 +1,73 @@
+def setup_environment():
+    """Fixture to set up the environment for Hugging Face downloads."""
+    # Check for Hugging Face token
+    token = os.getenv("HUGGINGFACE_TOKEN")
```
When we run the build on GitHub, the environment variable for HUGGINGFACE_TOKEN will not be set, so it will throw an error. If we would like to set, could we follow other unit tests for setting env variables?

rdheekonda commented on tests/test_hf_model_downloads.py:
```diff
@@ -0,0 +1,73 @@
+    yield  # Yield control back to the tests after setup
+
+
+def test_download_specific_files_with_aria2(setup_environment):
```
Could we update this unit test to check for behavior using pytest.raises, similar to the other exception tests in other test modules, and remove the print statement(s)?

rdheekonda commented on tests/test_hf_model_downloads.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def test_download_specific_files_with_cli(setup_environment):
+    """Test downloading specific files using huggingface-cli and measure time."""
```
Could we update this unit test to check for behavior using pytest.raises, similar to the other exception tests, and remove the print statement?

rdheekonda commented on tests/test_hf_model_downloads.py:
```diff
@@ -0,0 +1,73 @@
+
+
+def test_download_model_with_transformers(setup_environment):
+    """Test downloading model using transformers library and measure time."""
```
Could we update this unit test to check for behavior using pytest.raises, similar to the other exception tests, and remove the print statement?

KutalVolkan commented:
Hello @rdheekonda,

Thank you for your feedback, I really appreciate it! Please feel free to review again and suggest any new enhancements if needed. 

Regarding the model for the demo, `HuggingFaceTB/SmolLM-135M-Instruct,` I am not satisfied with the output. It doesn’t seem to answer the prompt correctly, even when I set the `max_token` to 100. The model I’ve found to work best is `microsoft/Phi-3-mini-4k-instruct.`

Perhaps the demo is intended primarily for inference time testing.

rdheekonda commented on tests/test_huggingface_chat_target.py:
```diff
@@ -0,0 +1,168 @@
+    )
+    prompt_request = PromptRequestResponse(request_pieces=[request_piece])
+
+    with pytest.raises(ValueError) as excinfo:
```
Could we consider simplifying this code because we no longer need to manually create and manage an event loop. 

Since pytest now natively supports async tests, we could simply declare the test function as `async` and use `await` to handle the asynchronous operations. For reference, could you follow the unit test from `tests\target\test_azure_openai_gptv_chat_target.py::test_send_prompt_async`. Here's how skeleton refactor code looks like:

```python
@pytest.mark.asyncio
async def test_send_prompt_async():
    await hf_chat.send_prompt_async(prompt_request=prompt_request)
```

rdheekonda replied:
This is not addressed, could you address this comment?

rdheekonda commented on pyrit/prompt_target/hugging_face_enpoint_target.py:
```diff
@@ -0,0 +1,129 @@
+
+        logger.info(f"Sending the following prompt to the cloud endpoint: {request.converted_value}")
+
+        async with httpx.AsyncClient() as client:
+            try:
+                response = await client.post(self.endpoint, headers=headers, json=payload)
```
Could we consider using `common\net_utility.make_request_and_raise_if_error_async` instead of implementing using HTTPX package directly. This approach is already unit-tested and will help maintain cleaner, more consistent code.

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,274 @@
+        if prompt_request.request_pieces[0].converted_value_data_type != "text":
+            raise ValueError("This target only supports text prompt input.")
+
+    def set_system_prompt(
+        self,
+        *,
+        system_prompt: str,
+        conversation_id: str,
+        orchestrator_identifier: Optional[dict[str, str]] = None,
+        labels: Optional[dict[str, str]] = None,
+    ) -> None:
+        """
+        Sets the system prompt for the conversation.
+        """
+        super().set_system_prompt(
+            system_prompt=system_prompt,
+            conversation_id=conversation_id,
+            orchestrator_identifier=orchestrator_identifier,
+            labels=labels,
+        )
```
We can remove this function because the `PromptChatTarget` class allows users to call it directly. When this class is invoked by orchestrators, it will set the system prompt automatically. For example, you can refer to `AzureOpenAIGPT4OChatTarget` as a reference.

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,274 @@
+        self,
+        *,
+        model_id: str,
+        api_key: Optional[str] = None,
```
Nit: consider function arguments name change that reflects HF access token; something like `hf_access_token`?

rdheekonda commented on pyrit/prompt_target/hugging_face_chat_target.py:
```diff
@@ -0,0 +1,274 @@
+    _cache_enabled = False
+
+    # Define the environment variable name for the Hugging Face token
+    API_KEY_ENVIRONMENT_VARIABLE = "HUGGINGFACE_TOKEN"
```
Could you add this environment variable name to `.env_example` file so users could follow setting up in their `.env`? Also, could you consider renaming the var name that matches to the HF token something like `HUGGINGFACE_TOKEN_ENVIRONMENT_VARIABLE`?

rdheekonda commented:
> Hello @rdheekonda,
> 
> Thank you for your feedback, I really appreciate it! Please feel free to review again and suggest any new enhancements if needed.
> 
> Regarding the model for the demo, `HuggingFaceTB/SmolLM-135M-Instruct,` I am not satisfied with the output. It doesn’t seem to answer the prompt correctly, even when I set the `max_token` to 100. The model I’ve found to work best is `microsoft/Phi-3-mini-4k-instruct.`
> 
> Perhaps the demo is intended primarily for inference time testing.


Thanks, @KutalVolkan, for addressing the comments. I’ve reviewed and added a few more comments. Let me know once they’re resolved, and it will be good to merge.

KutalVolkan commented:
> Thanks, @KutalVolkan, for addressing the comments. I’ve reviewed and added a few more comments. Let me know once they’re resolved, and it will be good to merge.

Hello Raja,

Thanks for the additional feedback! I've addressed the new comments. Please review and let me know if anything else is needed.

rdheekonda reviewed (approved):
Looks great, few things before merging

1. One comment needs to be addressed.
2. Merge onto main

KutalVolkan commented:
> Looks great, few things before merging
> 
> 1. One comment needs to be addressed.
> 2. Merge onto main

Hello Raja,

I hope I have made the changes you requested. If anything is still not quite right, I might need a bit more clarification. Please let me know if any further adjustments are needed.

Thank you for your support :)

rdheekonda commented:
> > Looks great, few things before merging
> > 
> > 1. One comment needs to be addressed.
> > 2. Merge onto main
> 
> Hello Raja,
> 
> I hope I have made the changes you requested. If anything is still not quite right, I might need a bit more clarification. Please let me know if any further adjustments are needed.
> 
> Thank you for your support :)

It's been addressed in the latest commit https://github.com/Azure/PyRIT/pull/354/commits/44b9251575ce55f000bc142456e29306ff377605
Thanks.

rdheekonda commented:
Upon running build, unit tests are failing with `torch` module not found in the environment. Could you add necessary dependencies to the `pyproject.yaml` so that we retry running the build?

KutalVolkan commented:
> Upon running build, unit tests are failing with `torch` module not found in the environment. Could you add necessary dependencies to the `pyproject.yaml` so that we retry running the build?

Hello Raja,

The dependency issue has been addressed. Thank you for pointing it out. If anything else needs to be done, I'm happy to make further adjustments.

romanlutz commented:
I don't see any changes to pyproject.toml. Am I missing something?

KutalVolkan commented:
> I don't see any changes to pyproject.toml. Am I missing something?

 Is what I did correct? I thought i addressed it here 14f6a3f

KutalVolkan commented:
Hello Roman & Raja,

I noticed the error regarding the Hugging Face token. I will handle that by tomorrow by mocking the token in the tests to bypass the need for a real `HUGGINGFACE_TOKEN`.

rdheekonda commented:
> Hello Roman & Raja,
> 
> I see the error regarding the Hugging Face token. I will handle that by tomorrow by mocking the token.

Thanks, @KutalVolkan. Here's the reference [code](https://github.com/Azure/PyRIT/blob/c7a9f08153e222b00486e2d9a28886d22542adef/tests/target/test_azure_openai_gptv_chat_target.py#L96C1-L104C10) for how we mocked the environment variables in our pytests.

KutalVolkan commented:
Hello @romanlutz & @rdheekonda,

I’ve updated the tests for `HuggingFaceChatTarget` to mock `get_required_value`. I’ve also added a test to ensure a `ValueError` is raised when the token is missing, thanks to Raja’s code snippet reference.

Fingers crossed this works as expected! 😊

romanlutz commented:
> > I don't see any changes to pyproject.toml. Am I missing something?
> 
> 
> 
>  Is what I did correct? I thought i addressed it here 14f6a3f

Somehow it wasn't showing before. This is fine.

rdheekonda commented:
Merged the PR. Excellent work, @KutalVolkan !

KutalVolkan commented:
> Merged the PR. Excellent work, @KutalVolkan !

Thank you, Raja and Roman, for your support!
